src/modules/types.rs

```diff
@@ -1,26 +1,108 @@
 use std::fmt::Display;
 
 use heraclitus_compiler::prelude::*;
+use itertools::Itertools;
 use crate::utils::ParserMetadata;
 
-#[derive(Debug, Clone, PartialEq, Eq, Default)]
+#[derive(Debug, Clone, Eq, Default)]
 pub enum Type {
     #[default] Null,
     Text,
     Bool,
     Num,
+    Union(Vec<Type>),
     Array(Box<Type>),
     Failable(Box<Type>),
     Generic
 }
 
+impl Type {
+    fn eq_union_normal(one: &[Type], other: &Type) -> bool {
+        one.iter().any(|x| (*x).to_string() == other.to_string())
+    }
+
+    fn eq_unions(one: &[Type], other: &[Type]) -> bool {
+        let bigger;
+        let smaller = if one.len() < other.len() { bigger = other; one } else { bigger = one; other };
+
+        smaller.iter().all(|x| {
+            Self::eq_union_normal(bigger, x)
+        })
+    }
+
+    /**
+        Hash is calculated in a hex number with 8 digits. Each digit has its own meaning.
+
+        Last two digits are reserved for "singular" types, like text and/or bool.
+
+        # Important
+        This hash is not supposed to be readed.  
+        It is generated in a way that it can't collide, but it is not supposed to be used for representing a type.
+
+        ```
+        0x00 00 00 00
+          ^^ ^^ ^^ ^^ -- singular type indicator
+                || -- -- number of nested arrays, such as [Num] will be 1, and [[[Num]]] will be 3.
+             || --    -- modifier
+          || --       -- reserved for future use
+        ```
+
+        ## Modifiers
+        These modifiers are valid:
+
+        | code | desc          |
+        | ---- | ------------- |
+        | `00` | no modifier   |
+        | `01` | failable type |
+     */
+    fn type_hash(&self) -> u32 {
+        match self {
+            Type::Null    => 0x00000001,
+            Type::Text    => 0x00000002,
+            Type::Bool    => 0x00000003,
+            Type::Num     => 0x00000004,
+            Type::Generic => 0x00000005,
+
+            Type::Array(t) => t.type_hash() + 0x00000100,
+
+            Type::Failable(t) => {
+                if let Type::Failable(_) = **t {
+                    panic!("Failable types can't be nested!");
+                }
+                t.type_hash() + 0x00010000
+            },
+
+            Type::Union(_) => unreachable!("Type hash is not available for union types! Use the PartialEq trait instead"),
+        }
+    }
+}
```

Comment:
I think this needs some proof, but it's late and I'm lazy XD

Try by doing some tests, or math. Or try with hexes to be sure
```suggestion
    fn type_hash(&self) -> u32 {
        match self {
            Type::Null    => 1 << 0,
            Type::Text    => 1 << 1,
            Type::Bool    => 1 << 2,
            Type::Num     => 1 << 3,
            Type::Generic => 1 << 4,
            Type::Array(t) => t.type_hash() + (1 << 5),

            Type::Failable(t) => {
                if let Type::Failable(_) = **t {
                    panic!("Failable types can't be nested!");
                }
                t.type_hash() +(1 << 6)
            },

            Type::Union(types) => sum(dedup(types)) + (1 << 7),
        }
    }
}
```

Reply:
i removed that function the same day i wrote it, you should review on the latest commit

Reply:
nvm, i forgot to push it.. lmao

Reply:
This function is pretty stinky though. What's the point of making these hashes? `PartialEq` isn't enough? `matches` macro is not good enough?